fearless/ApplicationLayer/Services/WalletBalanceSubscription/WalletBalanceBuilder.swift

```diff
@@ -0,0 +1,190 @@
+                for: chainAsset,
+                accountInfo,
+                prices
+            )
+
+            if accountInfos.keys.contains(chainAssetKey) {
+                accountInfosCount += 1
+            }
+
+            totalBalance += balance.total
+            totalDayChange += balance.dayChange
+        }
+
+        let isLoaded = accountInfosCount == chainAssets.count
+        return CountBalanceInfo(
+            totalBalance: totalBalance,
+            totalDayChange: totalDayChange,
+            isLoaded: isLoaded
+        )
+    }
+
+    private func split(
+        _ chainAssets: [ChainAsset],
+        for metaAccount: MetaAccountModel
+    ) -> (enabled: [ChainAsset], disabled: [ChainAsset]) {
+        var enabledChainAssets: [ChainAsset] = []
+        var disabledChainAssets: [ChainAsset] = []
+
+        chainAssets.forEach { chainAsset in
+            let accountRequest = chainAsset.chain.accountRequest()
+            guard let accountId = metaAccount.fetch(for: accountRequest)?.accountId else {
+                return
+            }
+            let chainAssetKey = chainAsset.uniqueKey(accountId: accountId)
+
+            if
+                let assetIdsEnabled = metaAccount.assetIdsEnabled,
+                assetIdsEnabled.contains(chainAssetKey) {
+                enabledChainAssets.append(chainAsset)
+            } else {
+                disabledChainAssets.append(chainAsset)
+            }
+        }
+
+        return (enabled: enabledChainAssets, disabled: disabledChainAssets)
+    }
+
+    private func getFiatBalance(
+        for chainAsset: ChainAsset,
+        _ accountInfo: AccountInfo?,
+        _ prices: [PriceData]
+    ) -> DayFiatBalance {
+        let balanceDecimal = getBalance(
+            for: chainAsset,
+            accountInfo
+        )
+
+        guard let priceId = chainAsset.asset.priceId,
+              let priceData = prices.first(where: { $0.priceId == priceId }),
+              let priceDecimal = Decimal(string: priceData.price)
+        else {
+            return DayFiatBalance(total: .zero, dayChange: .zero)
+        }
+
+        let total = priceDecimal * balanceDecimal
+        let dayChange = total * (priceData.fiatDayChange ?? .zero)
+        let dayFiatBalance = DayFiatBalance(total: total, dayChange: dayChange)
+
+        return dayFiatBalance
+    }
+
+    private func getBalance(
+        for chainAsset: ChainAsset,
+        _ accountInfo: AccountInfo?
+    ) -> Decimal {
+        guard
+            let accountInfo = accountInfo,
+            let balance = Decimal.fromSubstrateAmount(
+                accountInfo.data.total,
+                precision: chainAsset.asset.displayInfo.assetPrecision
+            )
+        else {
+            return .zero
+        }
+
+        return balance
+    }
+}
+
+private struct CountBalanceInfo {
+    let totalBalance: Decimal
+    let totalDayChange: Decimal
+    let isLoaded: Bool
+}
+
+private struct DayFiatBalance {
```

Comment:
don't we need to add currency parameter also ? you have decimal there, but it isn't clear what this number means ($, €, ₽).

Reply:
and also name of this struct should be AssetFiatBalanceInfo